### PR TITLE
More detail diagnostic when downcast to CoreFoundation Type

### PR DESCRIPTION
**Pull request by orakaro**
More detail diagnostic when downcast to CF type.
Resolves [SR-7015](https://bugs.swift.org/browse/SR-7015).

### CODE REVIEW

File: include/swift/AST/DiagnosticsSema.def [hunk 1]
```diff
@@ -2533,8 +2533,8 @@ NOTE(silence_inject_forced_downcast,none,
      "add parentheses around the cast to silence this warning", ())
 
 ERROR(conditional_downcast_foreign,none,
-      "conditional downcast to CoreFoundation type %0 will always succeed",
-      (Type))
+     "conditional downcast to CoreFoundation type %1 will always succeed; you would need to compare CFTypeID between %0 and %1",
```
Comment (CodaFi) [inline include/swift/AST/DiagnosticsSema.def]: This ought to be a note (and phrased in the active voice).

Comment (orakaro) [reply include/swift/AST/DiagnosticsSema.def]: @CodaFi Thanks for looking into this. Do you mean that the suggest part should be in separated NOTE ?

Comment (CodaFi) [reply include/swift/AST/DiagnosticsSema.def]: Yes

Comment (orakaro) [reply include/swift/AST/DiagnosticsSema.def]: Fixed! Please confirm.

File: include/swift/AST/DiagnosticsSema.def [hunk 2]
```diff
@@ -2533,8 +2533,11 @@ NOTE(silence_inject_forced_downcast,none,
      "add parentheses around the cast to silence this warning", ())
 
 ERROR(conditional_downcast_foreign,none,
-      "conditional downcast to CoreFoundation type %0 will always succeed",
+     "conditional downcast to CoreFoundation type %0 will always succeed",
```
Comment (CodaFi) [inline include/swift/AST/DiagnosticsSema.def]: Nit: Restore the whitespace here

File: include/swift/AST/DiagnosticsSema.def [hunk 3]
```diff
@@ -2533,8 +2533,11 @@ NOTE(silence_inject_forced_downcast,none,
...
       (Type))
+NOTE(compare_CFTyteId_foreign,none,
+     "compare CFTypeID between %0 and %1 to make sure they are convertible",
```
Comment (CodaFi) [inline include/swift/AST/DiagnosticsSema.def]: How about

> did you mean to explicitly compare the CFTypeIDs of %0 and %1?

Comment (CodaFi) [reply include/swift/AST/DiagnosticsSema.def]: Even this isnt necessarily a good wording for this.  The type of the cast variable is statically known to the type checker, but the type ID of the cast variable isn’t known until runtime.  %0 should mention the variable, not necessarily the type we inferred here.

File: include/swift/AST/DiagnosticsSema.def [hunk 4]
```diff
@@ -2533,8 +2533,11 @@ NOTE(silence_inject_forced_downcast,none,
      "add parentheses around the cast to silence this warning", ())
 
 ERROR(conditional_downcast_foreign,none,
-      "conditional downcast to CoreFoundation type %0 will always succeed",
+     "conditional downcast to CoreFoundation type %0 will always succeed",
       (Type))
+NOTE(compare_CFTyteId_foreign,none,
```
Comment (CodaFi) [inline include/swift/AST/DiagnosticsSema.def]: Nit: Could you fixup the selector’s name?

**Review by CodaFi — CHANGES_REQUESTED** [review]
Some nits and some wording stuff.  Looks good though!

File: include/swift/AST/DiagnosticsSema.def (hunk 2 above)
Comment (orakaro) [reply include/swift/AST/DiagnosticsSema.def]: 👌

File: include/swift/AST/DiagnosticsSema.def (hunk 4 above)
Comment (orakaro) [reply include/swift/AST/DiagnosticsSema.def]: 👌

File: include/swift/AST/DiagnosticsSema.def (hunk 3 above)
Comment (orakaro) [reply include/swift/AST/DiagnosticsSema.def]: Thanks! I am trying to get variable name for %0 here but not sure how to do it. Does it need to be an `Identifier` type? And how can I get it from `cast` or `sub` inside the method `getReferencedDecl`?

**Comment by orakaro** [comment]
I rewrote the note message. Please confirm @CodaFi 🙏

File: lib/Sema/CSApply.cpp [hunk 5]
```diff
@@ -3432,6 +3432,8 @@ namespace {
               auto &tc = cs.getTypeChecker();
               tc.diagnose(cast->getLoc(), diag::conditional_downcast_foreign,
                           destValueType);
+              tc.diagnose(cast->getLoc(), diag::note_explicitly_compare_cftypeid,
+                          sub->getReferencedDecl().getDecl()->getBaseName().getIdentifier(), destValueType);
```
Comment (CodaFi) [inline lib/Sema/CSApply.cpp]: This won't necessarily work for general expressions and will crash without a referenced declaration.  You should check if you have a reference declaration first.  This should also clean up this chain of accesses.

Comment (CodaFi) [reply lib/Sema/CSApply.cpp]: You can also put `DeclBaseName` in the diagnostic so you don't need to `getIdentifier()`.

File: lib/Sema/CSApply.cpp [hunk 6]
```diff
@@ -3432,6 +3432,11 @@ namespace {
               auto &tc = cs.getTypeChecker();
               tc.diagnose(cast->getLoc(), diag::conditional_downcast_foreign,
                           destValueType);
+              ConcreteDeclRef refDecl = sub->getReferencedDecl();
+              if (refDecl) {
```
Comment (CodaFi) [inline lib/Sema/CSApply.cpp]: Almost. You want to check `getDecl()` here.

Comment (orakaro) [reply lib/Sema/CSApply.cpp]: @CodaFi Thank you for guiding me till here. As I am still new to Cpp development I would like to clarify following things:
* Do we need to check the result of `getBaseName` to be not null as well?
* I found some code at [MiscDiagnostics](https://github.com/apple/swift/blob/master/lib/Sema/MiscDiagnostics.cpp#L1722) which is using a full chain inside diagnose method. Isn't it not safe too?

Comment (CodaFi) [reply lib/Sema/CSApply.cpp]: - `nullptr` is not an inhabitant of `DeclBaseName` 
- It is, but only if truly malformed AST manages to slip through all of Sema into CSApply, so it's morally correct.  My concern here is that you may not necessarily be diagnosing a declaration referenced by the cast target - you could very well have a pure expression in this position - so you will get `ConcreteDeclRef()` which has a null referenced declaration.

Comment (CodaFi) [reply lib/Sema/CSApply.cpp]: Ah, wait, there's an `operator bool()` here.

:shipit:

**Review by CodaFi — APPROVED** [review]

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by orakaro** [comment]
@CodaFi Thanks for explaining. When will this be merged?

**Comment by CodaFi** [comment]
Sorry for the delay, I honestly thought I merged this!

⛵️

**Comment by orakaro** [comment]
@CodaFi Sorry to bugging you after a long time. 

How can I know which version of Swift will have this change? Could you specify if there is a process to get the change actually included in new release?

**Comment by CodaFi** [comment]
This has been merged into the development snapshots as early as 5/10.  It's certainly been auto merged into swift 5.0, but my git-foo is not strong enough to check if you made it into an earlier release.

**Comment by orakaro** [comment]
Thank you very much!

**Pull request merged**